### PR TITLE
Xcode8 xm static registrar

### PR DESCRIPTION
**Pull request by chamons**
Also implements https://bugzilla.xamarin.com/show_bug.cgi?id=42889

### CODE REVIEW

**Comment by chamons** [comment]
@rolfbjarne Hacks make me sad, but they are effective. Do you have suggests on doing it better here? (Also please review).

**Comment by xamarin-release-manager** [comment]
Build failure

**Comment by chamons** [comment]
build

**Comment by xamarin-release-manager** [comment]
Build failure

File: tools/mmp/driver.cs [hunk 1]
```diff
@@ -998,6 +998,8 @@ static int Compile (IEnumerable<string> internalSymbols)
 					args.Append ("-g ");
 				args.Append ("-mmacosx-version-min=").Append (minos.ToString ()).Append (' ');
 				args.Append ("-arch ").Append (arch).Append (' ');
+				if (arch == "x86_64")
```
Comment (rolfbjarne) [inline tools/mmp/driver.cs]: This is unrelated to the other registrar changes, so could this go into a separate commit/PR?

File: tools/common/StaticRegistrar.cs [hunk 2]
```diff
@@ -3690,6 +3677,14 @@ void Generate (string header_path, string source_path)
 							hdr.WriteLine ("#include <objc/runtime.h>");
 							hdr.WriteLine ("#include <objc/message.h>");
 
+#if MONOMAC
+							// 32-bit header files do not define these types
+							if (!Is64Bits) {
```
Comment (rolfbjarne) [inline tools/common/StaticRegistrar.cs]: I think it makes more sense to check the availability attribute and see if each type is 64bit only instead of hardcoding specific types like this.

File: tools/mmp/driver.cs (hunk 1 above)
Comment (chamons) [inline tools/mmp/driver.cs]: Sure. I thought it was related enough but no problem.

File: tools/common/StaticRegistrar.cs (hunk 2 above)
Comment (chamons) [inline tools/common/StaticRegistrar.cs]: So the only reason I did these was to reduce the delta that could break customers (indirectly, was not sure if really could happen).

Before extensions, when we added those types, we would registers all types, including those 64-bit only on 32-bit builds. Adding the extension APIs broke this, since the types aren't even in the header files. Forward declaring these two let us register that mass of 64-bit only types even in 32-bit builds still..

Not registering 64-bit only types would obviously be cleaner, if you think it is safe @rolfbjarne. XM has the luxury of being less backwards compatible that XI.

**Comment by xamarin-release-manager** [comment]
Build failure

**Comment by chamons** [comment]
@rolfbjarne Re-review whenever you have time. I think I got it right now.

**Comment by xamarin-release-manager** [comment]
Build failure

**Comment by xamarin-release-manager** [comment]
Build failure

File: tools/common/StaticRegistrar.cs [hunk 3]
```diff
@@ -2228,15 +2230,32 @@ void Specialize (AutoIndentStringBuilder sb)
...
+					if (attributes != null && attributes.Any (x => x.Architecture == PlatformArchitecture.Arch64))
+						continue;
+				}
+
+				if (IsQTKitType (@class) && GetSDKVersion () >= new Version (10,12))
+					continue; // QTKit header was removed in 10.12 SDK
+
+				// These are 64-bit frameworks that extend NSExtensionContext / NSUserActivity, which you can't do
+				// if the header doesn't declare them. So hack it away, since they are useless in 64-bit anyway
+				if (!Is64Bits && (IsMapKitType (@class) || IsIntentsType (@class)))
```
Comment (rolfbjarne) [inline tools/common/StaticRegistrar.cs]: Is this still needed when you're skipping all the 64bit-only types above?

Comment (chamons) [inline tools/common/StaticRegistrar.cs]: Yes. I tried removing it. There must be another chunk of code that is handling the category code or something.

Even marking the category bits as onlyOn64 didn't fix it.

I did not look into it more as I got distracted with the extensions stuff. I can do more digging today if you desire.

Comment (rolfbjarne) [inline tools/common/StaticRegistrar.cs]: No, it's fine, I was just curious.
